### ISSUE TITLE
Out-of-core processing with dask not working properly?

### ISSUE THREAD

**Issue by lrntct** [pydata/xarray#2329]
Hi,

I have a bunch of GRIB files that amount to ~250GB. I want to concatenate them and save to zarr.
I concatenated them with CDO and saved to netcdf.
Now I have a ~500GB netcdf that I want to convert to zarr.
I want to convert to zarr because:
- I plan to run the analysis on a cluster, and I understand that zarr is better for that
- By using float16 and lz4 compression, I believe I can reduce the size  to ~100GB and have faster access (I think the analysis will be i/o bound).

The netcdf:
```
<xarray.Dataset>
Dimensions:  (lat: 721, lon: 1440, time: 119330)
Coordinates:
  * time     (time) datetime64[ns] 2000-01-01T06:00:00 2000-01-01T06:00:00 ...
  * lon      (lon) float32 0.0 0.25 0.5 0.75 1.0 1.25 1.5 1.75 2.0 2.25 2.5 ...
  * lat      (lat) float32 90.0 89.75 89.5 89.25 89.0 88.75 88.5 88.25 88.0 ...
Data variables:
    mtpr     (time, lat, lon) float32 ...
Attributes:
    CDI:          Climate Data Interface version ?? (http://mpimet.mpg.de/cdi)
    Conventions:  CF-1.6
    history:      Fri Jul 27 16:35:19 2018: cdo -f nc4 mergetime 2000-01.grib...
    institution:  European Centre for Medium-Range Weather Forecasts
    CDO:          Climate Data Operators version 1.9.3 (http://mpimet.mpg.de/...
```

The code:

```python
def netcdf2zarr(nc_path):
    chunks = {'time': -1, 'lat': 'auto', 'lon': 'auto'}
    encoding = {'mtpr': {'dtype': 'float16', 'compressor': zarr.Blosc(cname='lz4', clevel=9)},
                'lat': {'dtype': 'float32'},
                'lon': {'dtype': 'float32'}}
    ds = xr.open_dataset(nc_path).chunk(chunks)
    ds.to_zarr('myfile.zarr', encoding=encoding)
```

#### Problem description

I left my code to run over the weekend. After 63h of processing, the zarr store was only 1GB in size. The system monitor indicated that the Python process had 17TB worth of disk read.
At that rate, it would have taken months to finish.
Is there something that I can do to increase the processing speed?

I run Ubuntu 18.04 on a Core i7-6700 with 16GB of RAM. The disk is an HDD with a speed of ~100MB/s.

#### Output of ``xr.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 2.7.15.final.0
python-bits: 64
OS: Linux
OS-release: 4.15.0-24-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_GB.UTF-8
LOCALE: None.None

xarray: 0.10.8
pandas: 0.23.3
numpy: 1.14.2
scipy: 1.1.0
netCDF4: 1.3.1
h5netcdf: 0.6.1
h5py: 2.8.0
Nio: None
zarr: 2.2.0
bottleneck: 1.2.1
cyordereddict: 1.0.0
dask: 0.18.2
distributed: 1.22.0
matplotlib: 2.2.2
cartopy: 0.16.0
seaborn: 0.9.0
setuptools: 39.1.0
pip: 10.0.1
conda: None
pytest: None
IPython: None
sphinx: None

</details>

**Comment by rabernat** [comment]
@lrntct - this sounds like a reasonable way to use zarr. We routinely do this sort of transcoding and it works reasonable well. Unfortunately something clearly isn't working right in your case. These things can be hard to debug, but we will try to help you.

You might want to start by reviewing the [guide I wrote for Pangeo on preparing zarr datasets](http://pangeo-data.org/data.html#guide-to-preparing-cloud-optimized-data).

It would also be good to see a bit more detail. You posted a function `netcdf2zarr` that converts a single netcdf file to a single zarr file. How are you invoking that function? Are you trying to create one zarr store for each netCDF file? How many netCDF files are there? If there are many (e.g. one per timmestep), my recommendation is to create only one zarr store for the whole dataset. Open the netcdf files using `open_mfdataset`.

If instead you have just **one big netCDF file** as in the example you posted above, I think I see you problem: you are calling `.chunk()` after calling `open_dataset()`, rather calling `open_dataset(nc_path, chunks=chunks)`. This probably means that you are loading the whole dataset in a single task and then re-chunking. That could be the source of the inefficiency.

More ideas:
- explicitly specify the chunks (rather than using  `'auto'`)
- eliminate the negative number in your chunk sizes
- make sure you really need `clevel=9`


Another useful piece of advice would be to use the [dask distributed dashboard](http://distributed.readthedocs.io/en/latest/web.html) to monitor what is happening under the hood. You can do this by running
```python
from dask.distributed import Client
client = Client()
client
```
In a notebook, this should provide you a link to the scheduler dashboard. Once you call `ds.to_zarr()`, watch the task stream in the dashboard to see what is happening.

Hopefully these ideas can help you move forward.

**Comment by lrntct** [comment]
@rabernat Thanks for your answer.

I have one big NetCDF of ~500GB.
What I have changed:
- Run in a Jupyter notebook with distributed to get the dashboard
- Change the chunks to `{'lat': 90, 'lon': 90}`. That should be around 1GB per chunk.
- Chunk from the beginning with `ds = xr.open_dataset('my_netcdf.nc', chunks=chunks)`
- About the LZ4 compression, I did some test with a 1.5GB extract and the writing time was just 2% slower than uncompressed.

Now when I run `to_zarr()`, it creates a zarr store (~40kB) and all the workers start to read the disk, but they don't write anything.

The Dask dashboard looks like this:

![screenshot from 2018-07-30 of the Dask dashboard](https://user-images.githubusercontent.com/12278765/43405382-98e24ad2-9411-11e8-9526-a39e6204056a.png)

Is this the expected behaviour? I was somehow expecting that each worker will read a chunk and then write it to zarr, streamlined. This does not seem to be the case.

**Comment by rabernat** [comment]
Can you forget about zarr for a moment and just do a reduction on your dataset? For example:
```python
ds.sum().load()
```

Keep the same chunk arguments you are currently using. This will help us understand if the problem is with reading the files.

Is it you intention to chunk the files contiguously in time? Depending on the underlying structure of the data within the netCDF file, this could amount to a complete transposition of the data, which could be very slow / expensive. This could have some parallels with #2004.

**Comment by lrntct** [comment]
@rabernat I tried to do the sum. I have the same issue. The process just seems to read the disk endlessly, without even writing to the RAM:

![screenshot from 2018-07-31 10-05-04](https://user-images.githubusercontent.com/12278765/43450682-28db1f30-94ab-11e8-87f5-2d8cda915a04.png)

I tried to lower the chunk size, but it doesn't seem to change anything. Without chunk, I logically get a `MemoryError`.

I plan to do time-series analysis, so I thought that having contiguous chunks in time would be more efficient. The netcdf was created with `cdo -f nc4 mergetime`, so it should have mostly the same structure, I guess.

**Comment by fmaussion** [comment]
Out of curiosity: 
- why do you chunk over lats and lons rather than time? The order of dimensions in your dataarray suggest that chunking over time could be more efficient
- can you show the output of ``ds.mtpr`` and ``ds.mtpr.encoding`` ?

**Comment by fmaussion** [comment]
forget my comment about chunks - I thought this would make a difference but it's actually the opposite (to my surprise):

```python
import dask as da
import xarray as xr

da = xr.DataArray(da.array.zeros((3000, 721, 1440), chunks=(10, 721, 1440)), 
                             dims=('z', 'y', 'x'))
da.to_netcdf('da.nc')  # creates a 25GB file!

with xr.open_dataarray('da.nc', chunks={'z':10}) as da:
    %timeit da.sum()
5.55 ms ± 29.5 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

with xr.open_dataarray('da.nc', chunks={'y':10}) as da:
    %timeit da.sum()
2.05 ms ± 19.4 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

with xr.open_dataarray('da.nc', chunks={'x':10}) as da:
    %timeit da.sum()
3.03 ms ± 74.6 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

```

**Comment by lrntct** [comment]
@fmaussion `ds.mtpr`:
```
<xarray.DataArray 'mtpr' (time: 119330, lat: 721, lon: 1440)>
dask.array<shape=(119330, 721, 1440), dtype=float32, chunksize=(119330, 16, 16)>
Coordinates:
  * time     (time) datetime64[ns] 2000-01-01T06:00:00 2000-01-01T06:00:00 ...
  * lon      (lon) float32 0.0 0.25 0.5 0.75 1.0 1.25 1.5 1.75 2.0 2.25 2.5 ...
  * lat      (lat) float32 90.0 89.75 89.5 89.25 89.0 88.75 88.5 88.25 88.0 ...
Attributes:
    long_name:  Mean total precipitation rate
    units:      kg m**-2 s**-1
    code:       55
    table:      235
```
`ds.mtpr.encoding`:
```
{'complevel': 0, 'shuffle': False, 'dtype': dtype('float32'), 'contiguous': False,
 'zlib': False, 'source': u'my_netcdf.nc', 'fletcher32': False,
 'original_shape': (119330, 721, 1440), 'chunksizes': (1, 721, 1440)}
```
Those chunksizes are the opposite of what I was expecting...

**Comment by fmaussion** [comment]
> Those chunksizes are the opposite of what I was expecting...

`chunksizes` in `encoding` are ignored in your case, dask still uses your user provided encoding.

Can you still try to chunk along one dimension only? i.e. ``chunk={'time':1}`` or ``chunk={'lon':1}``

**Comment by fmaussion** [comment]
Sorry for the confusion, I had an obvious mistake in my timing experiment above (forgot to do the actual computations...).

The dimension order *does* make a difference:

```python
import dask as da
import xarray as xr

with xr.open_dataarray('da.nc', chunks={'z':10}) as d:
    %timeit d.sum().load()
3.94 s ± 95.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

with xr.open_dataarray('da.nc', chunks={'y':10}) as d:
    %timeit d.sum().load()
4.15 s ± 316 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

**Comment by lrntct** [comment]
I did some tests with my big netcdf. The chunking indeed makes a difference.

```
chunks = {'time': 'auto', 'lat': 'auto', 'lon': 'auto'}
ds = xr.open_dataset('era5_precip.nc', chunks=chunks)
ds.sum().load()

real	161m37.119s
user	33m9.720s
sys	63m47.696s

chunks = {'time': 1}
ds = xr.open_dataset('era5_precip.nc', chunks=chunks)
print(ds.sum().load())

real	109m55.839s
user	303m40.665s
sys	451m30.788s
```

I'll do some more tests with the calculation of the mean on the time axis, it might be more representative of what I want to do.

**Comment by lrntct** [comment]
I ran a comparison of the impact of chunk sizes with a profiler:

```python
profiler = Profiler()
for chunks in [{'time': 30}, {'lat': 30}, {'lon': 30}]:
    print(chunks)
    profiler.start()
    with xr.open_dataset(nc_path, chunks=chunks) as ds:
        print(ds.mean(dim='time').load())
    profiler.stop()
    print(profiler.output_text(unicode=True, color=True))
```

I am not sure if the profiler results are useful:

```
    {'time': 30}
    <xarray.Dataset>
    Dimensions:  (lat: 721, lon: 1440)
    Coordinates:
      * lon      (lon) float32 0.0 0.25 0.5 0.75 1.0 1.25 1.5 1.75 2.0 2.25 2.5 ...
      * lat      (lat) float32 90.0 89.75 89.5 89.25 89.0 88.75 88.5 88.25 88.0 ...
    Data variables:
        mtpr     (lat, lon) float32 8.30159e-06 8.30159e-06 8.30159e-06 ...
    5652.770 compare_chunks  read_grib.py:281
    └─ 5652.613 load  xarray/core/dataset.py:466
       └─ 5652.613 compute  dask/base.py:349
          └─ 5652.404 get  dask/threaded.py:33
             └─ 5652.400 get_async  dask/local.py:389
                └─ 5629.663 queue_get  dask/local.py:127
                   └─ 5629.663 get  Queue.py:150
                      └─ 5629.656 wait  threading.py:309
```

In the case of chunks on `lat` or `lon` only, I get a `MemoryError`.

I don't know if this helps, but it would be great to have a solution or workaround for that. Surely I am not the only one working with dataset of that size? What would be the best practice in my case?

**Comment by lrntct** [comment]
The dask task graph seems right (mean on the time dimension, lower number of chunks to make the visualisation practical):

![mean](https://user-images.githubusercontent.com/12278765/44461604-70ab6f80-a609-11e8-8213-4e6732b0a23f.png)

If I understand well, the 'getter' are doing the actual reading of the file, but in reality, they do not seem to run in parallel.

As for the zarr writing part, I do not know how to check the task graph.

**Comment by shoyer** [comment]
This might be worth testing with the changes from https://github.com/pydata/xarray/pull/2261, which refactors xarray's IO handling.

**Comment by lrntct** [comment]
It seems that I managed to get something working as it should.
I first load my monthly grib files with iris, convert to xarray, then write to zarr. This uses all the CPU cores, but loads the full array into memory. Since the individual arrays are relatively small, that is not an issue.
Then I load the monthly zarr stores with xarray, concatenate them with `auto_combine` and write to a big zarr. The memory usage peaked just above 17GB with 32 CPU threads. The array and chunks dimensions are:
`(time, latitude, longitude) float16 dask.array<shape=(113969, 721, 1440), chunksize=(113969, 20, 20)>`
I guess that reducing the chunk size will lower the memory usage.

Using that big zarr storage, plotting a map of the mean values along the time axis takes around 15min, uses all the cores and around 24GB of RAM.
The strange part is: I think I tried that before and it was not working[...](https://boyter.org/static/books/Cf7eHZ1W4AEeZJA.jpg)
